plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -128,70 +143,76 @@ sealed interface BinaryBooleanExpression : ExpEmbedding {
 data class And(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val pos: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.And(left.toViper(), right.toViper())
+    override fun toViper() = Exp.And(left.toViper(), right.toViper(), Position.KtSourcePosition(pos))
 }
 
 data class Or(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val pos: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.Or(left.toViper(), right.toViper())
+    override fun toViper() = Exp.Or(left.toViper(), right.toViper(), Position.KtSourcePosition(pos))
 }
 
 data class Implies(
     override val left: ExpEmbedding,
     override val right: ExpEmbedding,
+    override val pos: KtSourceElement? = null,
 ) : BinaryBooleanExpression {
-    override fun toViper() = Exp.Implies(left.toViper(), right.toViper())
+    override fun toViper() = Exp.Implies(left.toViper(), right.toViper(), Position.KtSourcePosition(pos))
 }
 
 data class Not(
     val exp: ExpEmbedding,
+    override val pos: KtSourceElement? = null,
 ) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
-    override fun toViper() = Exp.Not(exp.toViper())
+    override fun toViper() = Exp.Not(exp.toViper(), Position.KtSourcePosition(pos))
 }
 
 
 data object UnitLit : ExpEmbedding {
     override val type = UnitTypeEmbedding
+    override val pos: KtSourceElement? = null
 
     override fun toViper() = UnitDomain.element
 }
 
-data class IntLit(val value: Int) : ExpEmbedding {
+data class IntLit(val value: Int, override val pos: KtSourceElement? = null, ) : ExpEmbedding {
     override val type = IntTypeEmbedding
 
     override fun toViper() = Exp.IntLit(value)
 }
 
-data class BooleanLit(val value: Boolean) : ExpEmbedding {
+data class BooleanLit(val value: Boolean, override val pos: KtSourceElement? = null, ) : ExpEmbedding {
     override val type = BooleanTypeEmbedding
 
     override fun toViper() = Exp.BoolLit(value)
 }
 
 data class NullLit(val elemType: TypeEmbedding) : ExpEmbedding {
     override val type = NullableTypeEmbedding(elemType)
+    override val pos: KtSourceElement? = null
 
```

Comment:
I don't think it's accurate to say null literals never have positions; they do appear in constant expressions, for one.  I suspect the earlier implementation with `nullVal` made sense when we were using `Exp` directly, but now having this information would be more consistent.